providers/implementations/rands/fips_crng_test.c

```diff
@@ -0,0 +1,428 @@
+    crngt->apt.i = 1;
+    return 1;
+}
+
+static int crng_test(CRNG_TEST *crngt, const unsigned char *buf, size_t n)
+{
+    size_t i;
+
+    for (i = 0; i < n; i++)
+        if (!RCT_test(crngt, buf[i]) || !APT_test(crngt, buf[i])) {
+            crngt->state = EVP_RAND_STATE_ERROR;
+            ERR_raise(ERR_LIB_PROV,
+                      PROV_R_ENTROPY_SOURCE_FAILED_CONTINUOUS_TESTS);
+            return 0;
+        }
+    return 1;
+}
+
+static const OSSL_DISPATCH *find_call(const OSSL_DISPATCH *dispatch,
+                                      int function)
+{
+    if (dispatch != NULL)
+        while (dispatch->function_id != 0) {
+            if (dispatch->function_id == function)
+                return dispatch;
+            dispatch++;
+        }
+    return NULL;
+}
+
+static void *crng_test_new(void *provctx, void *parent,
+                           const OSSL_DISPATCH *p_dispatch)
+{
+    CRNG_TEST *crngt = OPENSSL_zalloc(sizeof(*crngt));
+    const OSSL_DISPATCH *pfunc;
+
+    if (crngt == NULL)
+        return NULL;
+
+    crngt->provctx = provctx;
+    crngt->state = EVP_RAND_STATE_UNINITIALISED;
+
+    /* Extract parent's functions */
+    if (parent != NULL) {
+        crngt->parent = parent;
+        if ((pfunc = find_call(p_dispatch, OSSL_FUNC_RAND_ENABLE_LOCKING)) != NULL)
+            crngt->parent_enable_locking = OSSL_FUNC_rand_enable_locking(pfunc);
+        if ((pfunc = find_call(p_dispatch, OSSL_FUNC_RAND_LOCK)) != NULL)
+            crngt->parent_lock = OSSL_FUNC_rand_lock(pfunc);
+        if ((pfunc = find_call(p_dispatch, OSSL_FUNC_RAND_UNLOCK)) != NULL)
+            crngt->parent_unlock = OSSL_FUNC_rand_unlock(pfunc);
+        if ((pfunc = find_call(p_dispatch, OSSL_FUNC_RAND_GETTABLE_CTX_PARAMS)) != NULL)
+            crngt->parent_gettable_ctx_params = OSSL_FUNC_rand_gettable_ctx_params(pfunc);
+        if ((pfunc = find_call(p_dispatch, OSSL_FUNC_RAND_GET_CTX_PARAMS)) != NULL)
+            crngt->parent_get_ctx_params = OSSL_FUNC_rand_get_ctx_params(pfunc);
+        if ((pfunc = find_call(p_dispatch, OSSL_FUNC_RAND_GET_SEED)) != NULL)
+            crngt->parent_get_seed = OSSL_FUNC_rand_get_seed(pfunc);
+        if ((pfunc = find_call(p_dispatch, OSSL_FUNC_RAND_CLEAR_SEED)) != NULL)
+            crngt->parent_clear_seed = OSSL_FUNC_rand_clear_seed(pfunc);
+    }
+
+    return crngt;
+}
+
+static void crng_test_free(void *vcrngt)
+{
+    CRNG_TEST *crngt = (CRNG_TEST *)vcrngt;
+
+    if (crngt != NULL) {
+        CRYPTO_THREAD_lock_free(crngt->lock);
+        OPENSSL_free(crngt);
+    }
+}
+
+static int crng_test_instantiate(void *vcrngt, unsigned int strength,
+                                 int prediction_resistance,
+                                 const unsigned char *pstr,
+                                 size_t pstr_len,
+                                 ossl_unused const OSSL_PARAM params[])
+{
+    CRNG_TEST *crngt = (CRNG_TEST *)vcrngt;
+
+    /* Start up health tests should go here */
```

Comment:
This sounds like there is something unfinished? Or should the comment be simply removed?

Reply:
They don't exist yet as far as I can tell.  The standard mentioned them but doesn't define any.

This comment was meant as a tag for the person who implements them when something is required.

Happy to remove but I think it's useful.

Reply:
OK then.

Reply:
I would either remove, or add to the comment with something similar too what you just said.